src/rest-server/src/models/v2/job-attempt.js

```diff
@@ -111,58 +96,43 @@ if (sequelize && launcherConfig.enabledJobHistory) {
   };
 
   const get = async (frameworkName, jobAttemptIndex) => {
-    let uid;
     let attemptFramework;
-    let response;
+    let framework;
+    const encodedFrameworkName = encodeName(frameworkName);
 
     try {
-      response = await k8sModel.getClient().get(
-        launcherConfig.frameworkPath(encodeName(frameworkName)),
-        {
-          headers: launcherConfig.requestHeaders,
-        }
+      framework = await databaseModel.Framework.findOne({
+        attributes: ['snapshot'],
+        where: {name: encodedFrameworkName}}
       );
     } catch (error) {
-      logger.error(`error when getting framework from k8s api: ${error.message}`);
-      if (error.response != null) {
-        response = error.response;
-      } else {
-        throw error;
-      }
+      logger.error(`error when getting framework from database: ${error.message}`);
+      throw error;
     }
 
-    if (response.status === 200) {
-      // get uid from k8s framwork API
-      uid = response.data.metadata.uid;
-      attemptFramework = response.data;
-    } else if (response.status === 404) {
-      logger.warn(`could not get framework ${uid} from k8s: ${JSON.stringify(response)}`);
-      return {status: 404, data: null};
+    if (framework) {
+      attemptFramework = JSON.parse(framework.snapshot);
     } else {
-      throw createError(response.status, 'UnknownError', response.data.message);
+      logger.warn(`could not get framework ${encodedFrameworkName} from database.`);
+      return {status: 404, data: null};
     }
 
     if (jobAttemptIndex < attemptFramework.spec.retryPolicy.maxRetryCount) {
-      if (isNil(uid)) {
-        return {status: 404, data: null};
-      }
-      const sqlSentence = `SELECT snapshot as data FROM framework_history WHERE ` +
-        `frameworkName = '${encodeName(frameworkName)}' and ` +
-        `attemptIndex = '${jobAttemptIndex}' ` +
-        `ORDER BY uid ASC;`;
-      const pgResult = (await sequelize.query(sqlSentence))[0];
+      const historyFramework = await databaseModel.FrameworkHistory.findOne({
+        attributes: ['snapshot'],
+        where: {frameworkName: encodedFrameworkName, attemptIndex: jobAttemptIndex},
+      });
```

Comment:
use the latest snapshot for the same framework and attemptIndex?

To judge if it is the  latest snapshot, could you also store the snapshot last write time in DB

Reply:
There is `insertedAt` and `updatedAt` field in history db.

What do you mean by `use the latest snapshot for the same framework and attemptIndex`?

Reply:
In the history DB, for the same framework name and attemptIndex, there may be many snapshots (some in running state, some in completed state), when show the retry history, we should give user the latest snapshot which is generally the completed state snapshot.

Reply:
Seems use updatedAt or snapshot last write time is not right. We should use the snapshot generation time instead of time write to db. If so, seems we should use `framework.status.transitionTime`

Reply:
For the same framework name and attemptIndex, there is **only one** record in the history db.

The history is recorded when fc log outputs something like `framework xxx is retried.`

Reply:
Will the history table store more snapshots in future? Such as rescale snapshot, or it is just retry snapshot?
BTW, even if it is just retry snapshot table, `framework xxx is retried.` may be produced by FC more than 1 time in some coner cases, will you override previous one, to make sure there is always only one?

Reply:
Currently only retry snapshots are available. For your corner case, there is no overwrite, two records will be all recorded.

Reply:
So you need to get the last one

Reply:
Use hash(frameworkName, attemptIndex, historyType) as primary key to solve this problem.